pkg/util/slowset.go

```diff
@@ -0,0 +1,94 @@
+/*
+Copyright 2024 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package util
+
+import (
+	"sync"
+	"time"
+)
+
+const (
+	slowSetSleepDelay = 100 * time.Millisecond
+)
+
+type SlowSet struct {
+	sync.RWMutex
+	// retentionTime is the time after which an item will be removed from the set
+	// this indicates, how long before an operation on pvc can be retried.
+	retentionTime time.Duration
+	workSet       map[string]time.Time
+}
+
+func NewSlowSet(retTime time.Duration) *SlowSet {
+	return &SlowSet{
+		retentionTime: retTime,
+		workSet:       make(map[string]time.Time),
+	}
+}
+
+func (s *SlowSet) Add(key string) bool {
+	s.Lock()
+	defer s.Unlock()
+
+	if _, ok := s.workSet[key]; ok {
+		return false
+	}
+
+	s.workSet[key] = time.Now()
+	return true
+}
+
+func (s *SlowSet) Contains(key string) bool {
+	s.RLock()
+	defer s.RUnlock()
+
+	_, ok := s.workSet[key]
+	return ok
+}
+
+func (s *SlowSet) Remove(key string) {
+	s.Lock()
+	defer s.Unlock()
+
+	delete(s.workSet, key)
+}
+
+func (s *SlowSet) TimeRemaining(key string) time.Duration {
+	s.RLock()
+	defer s.RUnlock()
+
+	if startTime, ok := s.workSet[key]; ok {
+		return s.retentionTime - time.Since(startTime)
+	}
+	return 0
+}
+
+func (s *SlowSet) Run(stopCh <-chan struct{}) {
+	for {
+		select {
+		case <-stopCh:
+			return
+		default:
+			time.Sleep(slowSetSleepDelay)
+			for key, t := range s.workSet {
+				if time.Since(t) > s.retentionTime {
+					s.Remove(key)
```

Comment:
Volumes should be removed also after successful ControllerExpand or after non-infeasible (= feasible) error.

Reply:
Hmm, originally I wrote this code to slow down expansion retries for failing node-expansions too. So something like:

1. Node expansion fails with infeasible error.
2. Controller retries, but adds the volume to `slowSet`.
3. Controller expansion succeeds and removes the volume from `slowSet`(*desired*)
4. Node expansion fails again with infeasible error.
5. Controller will immediately retry the expansion (*undesired*)

I agree though not removing volume from `slowSet` has a downside:

1. Controller expansion fails with infeasible error.
2. Controller retries and adds the volume to `slowSet`
3. Controller expansion somehow succeeds, but saving of one or more objects fail (because of resourceVersion check or something)
4. Now - controller will not correct this error until next 5 minutes (*undesired*)

A way to solve this would be perhaps to use - different sets for node and controller failures.

Reply:
Even if we solved this for controller expansion operations, it is much trickier to solve this for node expansion failures.

Reply:
okay, so I implemented removal via - https://github.com/kubernetes-csi/external-resizer/pull/418/commits/88252668b58ee62b35d8f3c1de75bb75774afecd